repl: add completion preview

BridgeAR opened the pull request:
This mainly adds the next step to our the newly added preview: (tab) completion preview. It is now also automatically inserted when moving the cursor right when being at the end of the current input instead of only when pushing tab.

See it in action:
https://asciinema.org/a/ePQx0GfCYQGdnQTzwlnSIyxbN

This also fixes at least two small issues:

1) ![image](https://user-images.githubusercontent.com/8822573/70650679-d6cd6300-1c4f-11ea-861d-bca7db02183a.png)
This is caused by the nested REPL tab completion. This was not detected earlier, since it's a rare case and only fails for syntax errors and similar. It was important to prevent syntax errors in the tab completion, otherwise the auto completion would have triggered those frequently.
2) It also fixes an issue with the former preview implementation where long lines could have caused similar REPL confusion.

Please have a look at the individual commits messages for further details.

// cc: @nodejs/repl 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

targos commented:
The asciinema link requires login.

BridgeAR commented:
@targos thanks, fixed.

targos commented on lib/internal/repl/utils.js:
```diff
@@ -28,6 +28,10 @@ const {
 
+const {
+  commonPrefix
+} = require('internal/readline/utils');
```
the variable is unused in this commit

BridgeAR replied:
Fixed

targos commented on lib/internal/readline/utils.js:
```diff
@@ -444,7 +445,27 @@ function* emitKeys(stream) {
   }
 }
 
+// This runs in O(n log n)
```
```suggestion
// This runs in O(n log n).
```

targos commented on lib/internal/repl/utils.js:
```diff
@@ -119,24 +123,98 @@ function isRecoverableError(e, code) {
+        return;
+      }
+
+      // Result and the text that was completed
```
```suggestion
      // Result and the text that was completed.
```

targos commented on lib/internal/repl/utils.js:
```diff
@@ -184,23 +262,51 @@ function setupPreview(repl, contextSymbol, bufferSymbol, active) {
+    // potential completion suffix. To do so, we also have to change the
+    // behavior of `enter` and `escape`:
+    // Enter should automatically add the suffix to the current line as long as
+    // escape was pressed. We might even remove the preview in case any cursor
```
```suggestion
    // escape was not pressed. We might even remove the preview in case any cursor
```

?

targos commented on lib/readline.js:
```diff
@@ -500,45 +501,47 @@ Interface.prototype._tabComplete = function(lastKeypressWasTab) {
-        }
-        handleGroup(self, group, width, maxColumns);
-      }
+    // Result and the text that was completed
```
```suggestion
    // Result and the text that was completed.
```

BridgeAR commented:
@targos I addressed your nits and added one more commit that improves the coverage significantly. This should cover almost all regular preview usage plus edge cases.

Trott commented on doc/api/repl.md:
```diff
@@ -565,8 +565,9 @@ changes:
-     **Default:** `true`. Always `false` in case `terminal` is falsy.
+  * `preview` {boolean} Defines if the repl prints autocomplete and output
+    previews or not. **Default:** `true`. Always `false` in case `terminal` is
+    falsy.
```
Does that mean the default is `true` if terminal is truthy and `false` if `terminal` is falsy? If that's the case then I think the text can be clarified a little, but that's a nit and not a blocking comment.

BridgeAR replied:
It's actually not only about the default. If the terminal is false, the preview is always deactivated. I am happy for suggestions to better word this :)

Trott replied:
Maybe instead of this

```markdown
Always `false` in case `terminal` is falsy.
```

...this?:

```markdown
If `terminal` is falsy, then there are no previews and the value of `preview` has no effect.
```

BridgeAR commented:
Rebased due to conflicts.

@Trott @targos I also pushed two very minor commits. There was a small bug that I still found and the test case that I thought would catch it was not strict enough before. Those two things are fixed in the extra commits. It would be great to get a LG confirmation to land this.

targos commented:
Lgtm

BridgeAR commented:
Landed in 67ed526ab03c3e9b2f41290b7cd6e9c2e7d9b4dd...ba29e270f3e8c86e45e7e323949d23955fccae1d :tada:
